### PR TITLE
t/400: Resolve infinite loop because of selection change

### PR DESCRIPTION
**Pull request by pjasiun**
Fixes: #400.

### CODE REVIEW

File: tests/view/observer/selectionobserver.js [hunk 1]
```diff
@@ -124,6 +130,71 @@ describe( 'SelectionObserver', () => {
...
+			done();
+
+			// Prevent "done() called multiple times" error.
+			done = () => {};
+		} );
+
+		changeDomSelection();
+	} );
+
+	it( 'should not be thread as an infinite loop if the position is different', ( done ) => {
```
Comment (scofalik) [inline tests/view/observer/selectionobserver.js]: "be treated"

File: tests/view/observer/selectionobserver.js [hunk 2]
```diff
@@ -124,6 +130,71 @@ describe( 'SelectionObserver', () => {
...
+				setTimeout( () => changeCollapsedDomSelection( counter ) );
+			} else {
+				done();
+			}
+		} );
+
+		changeCollapsedDomSelection( counter );
+	} );
+
+	it( 'should not be thread as an infinite loop if it is less then 3 times', ( done ) => {
```
Comment (scofalik) [inline tests/view/observer/selectionobserver.js]: "be treated"

**Comment by scofalik** [comment]
The solution looks a bit ugly but probably could not be better, especially if we want to have it soon. Still better than setting timeouts. For me it's okay, but I'll also wait for @reinmar to look at it as it is an important fix.

File: tests/view/observer/selectionobserver.js (hunk 2 above)
Comment (Reinmar) [inline tests/view/observer/selectionobserver.js]: And you don't need `()`.

File: src/view/observer/selectionobserver.js [hunk 3]
```diff
@@ -8,6 +8,8 @@
 import Observer from './observer.js';
 import MutationObserver from './mutationobserver.js';
 
+import log from '../../../utils/log.js';
```
Comment (Reinmar) [inline src/view/observer/selectionobserver.js]: I believe this is a first usage of this function :D.

File: tests/view/observer/selectionobserver.js (hunk 2 above)
Comment (scofalik) [inline tests/view/observer/selectionobserver.js]: Haven't we agreed to use `()` always, for better readability?

Comment (Reinmar) [inline tests/view/observer/selectionobserver.js]: Which reminds me that we haven't clarified this in https://github.com/ckeditor/ckeditor5-design/wiki/Code-Style. Let's pretend we haven't seen this.

**Comment by scofalik** [comment]
I'd also consider two things:
- getting rid of `_lastButOneSelection` and `_lastSelection` and just base on counter
- putting `_lastButOneSelection` and `_lastSelection` and probably `isInfiniteLoop` out of the class as they give really bad smell :P

**Comment by pjasiun** [comment]
> getting rid of _lastButOneSelection and _lastSelection and just base on counter

The problem is how to reset counter. I do not see any good event we could listen on to reset it.

> putting _lastButOneSelection and _lastSelection and probably isInfiniteLoop out of the class as they give really bad smell :P

Well, such hacks are always bad smells. All of these properties are private and used only in one function. They can not be local and I do not feel that making them global would make them smell better.

**Comment by Reinmar** [comment]
I added more comments and fixed some existing ones.

This is a hack of course, but it's quite elegant. It's enclosed in one mechanism, fully private and the solution is easy to grasp. So I'm ok with it.

**Pull request merged**